File: frontend/src/pages/Analytics.js
import React, { useEffect, useState } from 'react';
import { analyticsAPI } from '../utils/api';
import { BarChart3, TrendingUp, Target, Calendar, Award } from 'lucide-react';

function Analytics() {
  const [analytics, setAnalytics] = useState({
    overview: {},
    productivity: {},
    habits: {},
    streaks: {}
  });
  const [loading, setLoading] = useState(true);
  const [activeTab, setActiveTab] = useState('overview');

  useEffect(() => {
    loadAnalytics();
  }, []);

  const loadAnalytics = async () => {
    try {
      const [overviewRes, productivityRes, habitsRes, streaksRes] = await Promise.all([
        analyticsAPI.getOverview(),
        analyticsAPI.getProductivity(),
        analyticsAPI.getHabits(),
        analyticsAPI.getStreaks()
      ]);

      setAnalytics({
        overview: overviewRes.data,
        productivity: productivityRes.data,
        habits: habitsRes.data,
        streaks: streaksRes.data
      });
    } catch (error) {
      console.error('Error loading analytics:', error);
    } finally {
      setLoading(false);
    }
  };

  const tabs = [
    { id: 'overview', name: 'Overview', icon: BarChart3 },
    { id: 'productivity', name: 'Productivity', icon: TrendingUp },
    { id: 'habits', name: 'Habits', icon: Target },
    { id: 'streaks', name: 'Streaks', icon: Award },
  ];

  if (loading) {
    return (
      <div className="flex items-center justify-center h-64">
        <div className="animate-spin rounded-full h-8 w-8 border-b-2 border-primary-600"></div>
      </div>
    );
  }

  return (
    <div className="space-y-6">
      <div className="flex items-center justify-between">
        <h1 className="text-2xl font-bold text-secondary-900">Analytics</h1>
        <div className="text-sm text-secondary-500">
          Last updated: {new Date().toLocaleString()}
        </div>
      </div>

      {/* Tabs */}
      <div className="border-b border-secondary-200">
        <nav className="flex space-x-8">
          {tabs.map((tab) => (
            <button
              key={tab.id}
              onClick={() => setActiveTab(tab.id)}
              className={`py-2 px-1 border-b-2 font-medium text-sm ${
                activeTab === tab.id
                  ? 'border-primary-500 text-primary-600'
                  : 'border-transparent text-secondary-500 hover:text-secondary-700 hover:border-secondary-300'
              }`}
            >
              <div className="flex items-center space-x-2">
                <tab.icon className="h-4 w-4" />
                <span>{tab.name}</span>
              </div>
            </button>
          ))}
        </nav>
      </div>

      {/* Tab Content */}
      {activeTab === 'overview' && <OverviewTab data={analytics.overview} />}
      {activeTab === 'productivity' && <ProductivityTab data={analytics.productivity} />}
      {activeTab === 'habits' && <HabitsTab data={analytics.habits} />}
      {activeTab === 'streaks' && <StreaksTab data={analytics.streaks} />}
    </div>
  );
}

function OverviewTab({ data }) {
  return (
    <div className="space-y-6">
      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6">
        <MetricCard
          title="Total Tasks"
          value={data.total_tasks || 0}
          subtitle={`${Math.round(data.completion_rate || 0)}% completed`}
          color="primary"
        />
        <MetricCard
          title="This Week"
          value={data.completed_this_week || 0}
          subtitle={`${Math.round(data.weekly_completion_rate || 0)}% completion`}
          color="success"
        />
        <MetricCard
          title="Active Habits"
          value={data.active_habits || 0}
          subtitle={`${Math.round(data.habit_consistency || 0)}% consistency`}
          color="warning"
        />
        <MetricCard
          title="Overall Score"
          value={Math.round((data.completion_rate + data.habit_consistency) / 2) || 0}
          subtitle="Performance index"
          color="secondary"
        />
      </div>

      <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
        <div className="card">
          <h3 className="text-lg font-semibold text-secondary-900 mb-4">Progress Summary</h3>
          <div className="space-y-4">
            <ProgressBar
              label="Task Completion"
              value={data.completion_rate || 0}
              color="primary"
            />
            <ProgressBar
              label="Weekly Goals"
              value={data.weekly_completion_rate || 0}
              color="success"
            />
            <ProgressBar
              label="Habit Consistency"
              value={data.habit_consistency || 0}
              color="warning"
            />
          </div>
        </div>

        <div className="card">
          <h3 className="text-lg font-semibold text-secondary-900 mb-4">Quick Insights</h3>
          <div className="space-y-3">
            <InsightItem
              text={`You've completed ${data.completed_tasks || 0} out of ${data.total_tasks || 0} total tasks`}
              type="info"
            />
            <InsightItem
              text={`This week you completed ${data.completed_this_week || 0} tasks`}
              type="success"
            />
            <InsightItem
              text={`You're tracking ${data.active_habits || 0} active habits`}
              type="info"
            />
            {data.habit_consistency < 70 && (
              <InsightItem
                text="Consider focusing on habit consistency to improve overall performance"
                type="warning"
              />
            )}
          </div>
        </div>
      </div>
    </div>
  );
}

function ProductivityTab({ data }) {
  return (
    <div className="space-y-6">
      <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
        <div className="card">
          <h3 className="text-lg font-semibold text-secondary-900 mb-4">Time by Priority</h3>
          {data.time_by_priority && data.time_by_priority.length > 0 ? (
            <div className="space-y-3">
              {data.time_by_priority.map((item) => (
                <div key={item.priority} className="flex items-center justify-between">
                  <span className="text-secondary-600 capitalize">{item.priority} Priority</span>
                  <span className="font-medium text-secondary-900">{item.hours.toFixed(1)}h</span>
                </div>
              ))}
            </div>
          ) : (
            <p className="text-secondary-500 text-center py-8">No productivity data available yet</p>
          )}
        </div>

        <div className="card">
          <h3 className="text-lg font-semibold text-secondary-900 mb-4">Weekly Trend</h3>
          {data.weekly_trend && data.weekly_trend.length > 0 ? (
            <div className="space-y-2">
              {data.weekly_trend.map((week) => (
                <div key={week.week} className="flex items-center justify-between">
                  <span className="text-secondary-600">{week.week}</span>
                  <span className="font-medium text-secondary-900">{week.completed_tasks} tasks</span>
                </div>
              ))}
            </div>
          ) : (
            <p className="text-secondary-500 text-center py-8">No trend data available yet</p>
          )}
        </div>
      </div>
    </div>
  );
}

function HabitsTab({ data }) {
  return (
    <div className="space-y-6">
      <div className="card">
        <h3 className="text-lg font-semibold text-secondary-900 mb-4">Habit Performance (Last 30 Days)</h3>
        {data.habits && data.habits.length > 0 ? (
          <div className="space-y-4">
            {data.habits.map((habit) => (
              <div key={habit.habit_name} className="border border-secondary-200 rounded-lg p-4">
                <div className="flex items-center justify-between mb-2">
                  <h4 className="font-medium text-secondary-900">{habit.habit_name}</h4>
                  <span className="badge badge-secondary">{habit.category}</span>
                </div>
                <div className="grid grid-cols-2 md:grid-cols-4 gap-4 text-sm">
                  <div>
                    <span className="text-secondary-500">Target</span>
                    <p className="font-medium">{habit.target_frequency}x/week</p>
                  </div>
                  <div>
                    <span className="text-secondary-500">Completion</span>
                    <p className="font-medium">{Math.round(habit.completion_rate)}%</p>
                  </div>
                  <div>
                    <span className="text-secondary-500">Total Entries</span>
                    <p className="font-medium">{habit.total_entries}</p>
                  </div>
                  <div>
                    <span className="text-secondary-500">Completed</span>
                    <p className="font-medium">{habit.completed_entries}</p>
                  </div>
                </div>
                <ProgressBar
                  label=""
                  value={habit.completion_rate}
                  color="success"
                  className="mt-3"
                />
              </div>
            ))}
          </div>
        ) : (
          <p className="text-secondary-500 text-center py-8">No habit data available yet</p>
        )}
      </div>
    </div>
  );
}

function StreaksTab({ data }) {
  return (
    <div className="space-y-6">
      <div className="card">
        <h3 className="text-lg font-semibold text-secondary-900 mb-4">Current Streaks</h3>
        {data.streaks && data.streaks.length > 0 ? (
          <div className="space-y-4">
            {data.streaks.map((streak, index) => (
              <div key={index} className="flex items-center justify-between p-4 bg-secondary-50 rounded-lg">
                <div>
                  <h4 className="font-medium text-secondary-900">{streak.type}</h4>
                  <p className="text-sm text-secondary-600">{streak.description}</p>
                </div>
                <div className="text-right">
                  <p className="text-2xl font-bold text-primary-600">{streak.current_streak}</p>
                  <p className="text-sm text-secondary-500">days</p>
                </div>
              </div>
            ))}
          </div>
        ) : (
          <p className="text-secondary-500 text-center py-8">No streaks data available yet</p>
        )}
      </div>
    </div>
  );
}

function MetricCard({ title, value, subtitle, color }) {
  const colorClasses = {
    primary: 'text-primary-600',
    success: 'text-success-600',
    warning: 'text-warning-600',
    secondary: 'text-secondary-600',
  };

  return (
    <div className="card">
      <h3 className="text-sm font-medium text-secondary-600">{title}</h3>
      <p className={`text-2xl font-semibold mt-1 ${colorClasses[color]}`}>{value}</p>
      <p className="text-sm text-secondary-500 mt-1">{subtitle}</p>
    </div>
  );
}

function ProgressBar({ label, value, color, className = '' }) {
  const colorClasses = {
    primary: 'bg-primary-500',
    success: 'bg-success-500',
    warning: 'bg-warning-500',
    secondary: 'bg-secondary-500',
  };

  return (
    <div className={className}>
      {label && (
        <div className="flex justify-between text-sm mb-1">
          <span className="text-secondary-600">{label}</span>
          <span className="text-secondary-900 font-medium">{Math.round(value)}%</span>
        </div>
      )}
      <div className="bg-secondary-200 rounded-full h-2">
        <div 
          className={`h-2 rounded-full ${colorClasses[color]}`}
          style={{ width: `${Math.min(value, 100)}%` }}
        ></div>
      </div>
    </div>
  );
}

function InsightItem({ text, type }) {
  const typeClasses = {
    info: 'text-primary-700 bg-primary-50 border-primary-200',
    success: 'text-success-700 bg-success-50 border-success-200',
    warning: 'text-warning-700 bg-warning-50 border-warning-200',
  };

  return (
    <div className={`p-3 rounded-lg border text-sm ${typeClasses[type]}`}>
      {text}
    </div>
  );
}

export default Analytics;

File: backend/app/routers/analytics.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List
from datetime import datetime, timedelta
from ..models.database import get_db
from ..models.models import Task, Habit, HabitEntry, User
from .auth import get_current_user

router = APIRouter()

@router.get("/overview")
def get_analytics_overview(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """Get analytics overview for dashboard"""
    
    # Task completion stats
    total_tasks = db.query(Task).filter(Task.owner_id == current_user.id).count()
    completed_tasks = db.query(Task).filter(Task.owner_id == current_user.id, Task.is_completed == True).count()
    
    # This week's tasks
    week_start = datetime.now() - timedelta(days=7)
    tasks_this_week = db.query(Task).filter(
        Task.owner_id == current_user.id,
        Task.created_at >= week_start
    ).count()
    
    completed_this_week = db.query(Task).filter(
        Task.owner_id == current_user.id,
        Task.completed_at >= week_start,
        Task.is_completed == True
    ).count()
    
    # Habit consistency
    active_habits = db.query(Habit).filter(Habit.owner_id == current_user.id, Habit.is_active == True).count()
    
    # Recent habit entries (last 7 days)
    recent_entries = db.query(HabitEntry).join(Habit).filter(
        Habit.owner_id == current_user.id,
        HabitEntry.date >= week_start
    ).count()
    
    completed_entries = db.query(HabitEntry).join(Habit).filter(
        Habit.owner_id == current_user.id,
        HabitEntry.date >= week_start,
        HabitEntry.completed == True
    ).count()
    
    return {
        "total_tasks": total_tasks,
        "completed_tasks": completed_tasks,
        "completion_rate": (completed_tasks / total_tasks * 100) if total_tasks > 0 else 0,
        "tasks_this_week": tasks_this_week,
        "completed_this_week": completed_this_week,
        "weekly_completion_rate": (completed_this_week / tasks_this_week * 100) if tasks_this_week > 0 else 0,
        "active_habits": active_habits,
        "habit_consistency": (completed_entries / recent_entries * 100) if recent_entries > 0 else 0
    }

@router.get("/productivity")
def get_productivity_analytics(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """Get productivity analytics - time spent per category"""
    
    # Time spent analysis
    time_by_priority = db.query(
        Task.priority,
        func.sum(Task.actual_hours).label('total_hours')
    ).filter(
        Task.owner_id == current_user.id,
        Task.actual_hours.isnot(None)
    ).group_by(Task.priority).all()
    
    # Weekly completion trend (last 8 weeks)
    weekly_data = []
    for i in range(8):
        week_start = datetime.now() - timedelta(weeks=i+1)
        week_end = datetime.now() - timedelta(weeks=i)
        
        completed = db.query(Task).filter(
            Task.owner_id == current_user.id,
            Task.completed_at >= week_start,
            Task.completed_at < week_end,
            Task.is_completed == True
        ).count()
        
        weekly_data.append({
            "week": f"Week {8-i}",
            "completed_tasks": completed
        })
    
    return {
        "time_by_priority": [{"priority": item.priority, "hours": float(item.total_hours or 0)} for item in time_by_priority],
        "weekly_trend": list(reversed(weekly_data))
    }

@router.get("/habits")
def get_habit_analytics(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """Get habit tracking analytics"""
    
    # Habit completion rates
    habits_data = []
    user_habits = db.query(Habit).filter(Habit.owner_id == current_user.id, Habit.is_active == True).all()
    
    for habit in user_habits:
        # Last 30 days
        thirty_days_ago = datetime.now() - timedelta(days=30)
        
        total_entries = db.query(HabitEntry).filter(
            HabitEntry.habit_id == habit.id,
            HabitEntry.date >= thirty_days_ago
        ).count()
        
        completed_entries = db.query(HabitEntry).filter(
            HabitEntry.habit_id == habit.id,
            HabitEntry.date >= thirty_days_ago,
            HabitEntry.completed == True
        ).count()
        
        completion_rate = (completed_entries / total_entries * 100) if total_entries > 0 else 0
        
        habits_data.append({
            "habit_name": habit.name,
            "category": habit.category,
            "target_frequency": habit.target_frequency,
            "completion_rate": completion_rate,
            "total_entries": total_entries,
            "completed_entries": completed_entries
        })
    
    return {"habits": habits_data}

@router.get("/streaks")
def get_streaks(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """Calculate current streaks for habits and tasks"""
    
    streaks = []
    
    # Task completion streak
    task_streak = 0
    current_date = datetime.now().date()
    
    while True:
        day_tasks = db.query(Task).filter(
            Task.owner_id == current_user.id,
            func.date(Task.due_date) == current_date,
            Task.is_completed == True
        ).count()
        
        if day_tasks > 0:
            task_streak += 1
            current_date -= timedelta(days=1)
        else:
            break
    
    streaks.append({
        "type": "Daily Tasks",
        "current_streak": task_streak,
        "description": f"{task_streak} days of completing at least one task"
    })
    
    return {"streaks": streaks}